compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/TailrecLowering.kt

```diff
@@ -52,19 +53,29 @@ open class TailrecLowering(val context: BackendContext) : BodyLoweringPass {
 
                 override fun visitFunction(declaration: IrFunction) {
                     declaration.acceptChildrenVoid(this)
-                    lowerTailRecursionCalls(context, declaration, useProperComputationOrderOfTailrecDefaultParameters())
+                    lowerTailRecursionCalls(declaration)
                 }
             })
 
-            lowerTailRecursionCalls(context, container, useProperComputationOrderOfTailrecDefaultParameters())
+            lowerTailRecursionCalls(container)
         }
     }
 
+    private fun lowerTailRecursionCalls(function: IrFunction) =
+        lowerTailRecursionCalls(context, function, useProperComputationOrderOfTailrecDefaultParameters(), ::followFunctionReference)
+
     open fun useProperComputationOrderOfTailrecDefaultParameters() = true
+
+    open fun followFunctionReference(reference: IrFunctionReference): Boolean = false
```

Comment:
Why don't enable it by default?

Reply:
In JS_IR and Native this lowering should run after IR inlining already. The INLINE_LAMBDA origin is JVM-specific.

Reply:
So if there is no inline lambda left tailrec optimization should work automatically?

Reply:
Yes.